File: TaskFlow App/lib/core/network/api_response.dart
import 'package:json_annotation/json_annotation.dart';

part 'api_response.g.dart';

@JsonSerializable(genericArgumentFactories: true)
class ApiResponse<T> {
	final bool succeeded;
	final String? message;
	final List<String>? errors;
	final T? data;

	const ApiResponse({required this.succeeded, this.message, this.errors, this.data});

	factory ApiResponse.fromJson(Map<String, dynamic> json, T Function(Object?) fromJsonT) =>
			_$ApiResponseFromJson(json, fromJsonT);
}

@JsonSerializable(genericArgumentFactories: true)
class PagedResult<T> {
	final List<T> items;
	final int pageNumber;
	final int pageSize;
	final int totalPages;
	final int totalCount;
	final bool hasPreviousPage;
	final bool hasNextPage;
	final int page;

	const PagedResult({
		required this.items, required this.pageNumber, required this.pageSize,
		required this.totalPages, required this.totalCount,
		required this.hasPreviousPage, required this.hasNextPage, required this.page,
	});

	factory PagedResult.fromJson(Map<String, dynamic> json, T Function(Object?) fromJsonT) =>
			_$PagedResultFromJson(json, fromJsonT);
}

File: TaskFlow/src/TaskFlow.Shared/Wrappers/ApiResponse.cs
namespace TaskFlow.Shared.Wrappers;

public class ApiResponse<T>
{
    public bool Succeeded { get; set; }
    public string? Message { get; set; }
    public List<string>? Errors { get; set; }
    public T? Data { get; set; }

    public ApiResponse() { }

    public ApiResponse(T data, string? message = null)
    {
        Succeeded = true;
        Message = message;
        Data = data;
    }

    public static ApiResponse<T> Success(T data, string? message = null) => new(data, message);
    public static ApiResponse<T> Ok(T data, string? message = null) => new(data, message);
    public static ApiResponse<T> Fail(string message) => new() { Succeeded = false, Message = message };
    public static ApiResponse<T> Fail(List<string> errors) => new() { Succeeded = false, Errors = errors };
}
